packages/gatsby-plugin-benchmarks/gatsby-node.js

```diff
@@ -0,0 +1,124 @@
+const nodeFetch = require(`node-fetch`)
+
+const { performance } = require(`perf_hooks`)
+
+const bootstrapTime = performance.now()
+
+let apiServer
+let apiPath
+
+class BenchMeta {
+  constructor() {
+    this.flushing = false // Started flushing?
+    this.flushed = false // Completed flushing?
+    this.crashed = false
+    this.localTime = new Date().toJSON()
+    this.events = {
+      // TODO: we should also have access to node's timing data and see how long it took before bootstrapping this script
+      bootstrapTime,
+      instanceTime: performance.now(),
+      start: 0,
+      bootstrap: 0,
+      stop: 0,
+    }
+    this.started = false
+  }
+
+  markStart() {
+    if (this.started) {
+      this.crashed = "Error: Should not call markStart() more than once"
+      console.error("gatsby-plugin-benchmarks: " + this.crashed)
+      TODO // report error.
+    }
+    this.events.start = performance.now()
+    this.started = true
+  }
+
+  markDataPoint(name) {
+    this.events[name] = performance.now()
+  }
+
+  async markStop() {
+    if (!this.events.start) {
+      this.crashed =
+        "Error: Should not call markStop() before calling markStart()"
+      console.error("gatsby-plugin-benchmarks: " + this.crashed)
+      TODO // report error.
+    }
+    this.events.stop = performance.now()
+    return this.flush()
+  }
+
+  async flush() {
+    this.flushing = true
+
+    return nodeFetch(`https://${apiServer}${apiPath}`, {
+      method: `POST`,
+      headers: {
+        "content-type": `application/json`,
+        // "user-agent": this.getUserAgent(),
+      },
+      body: JSON.stringify({
+        time: this.localTime,
+        // This is a stub. We'll expand this as we determine what to log when running a benchmark
+        sessionId: JSON.stringify(this.events),
+      }),
+    })
+      .then(res => {
+        this.flushed = true
+        // Note: res.text returns a promise
+        return res.text()
+      })
+      .then(text => console.log("Server response:", text))
+  }
+}
+
+process.on(`exit`, async () => {
+  if (!benchMeta.crashed) {
+    // This should not happen (benchmarks are controlled) and have been reported already
+  } else if (!benchMeta.flushing) {
+    console.log(
+      "gatsby-plugin-benchmarks: This is process.exit(); Not yet flushed, will flush now but it's probably too late..."
+    )
+    benchMeta.markDataPoint("post-build")
+    return benchMeta.markStop()
+  } else if (!benchMeta.flushed) {
+    // Started to flush but the completion promise did not fire yet
+    // This should't happen unless the reporting crashed hard
+  }
+})
+
+const benchMeta = new BenchMeta()
+
+async function onPreInit(api, options) {
+  // console.log("# onPreInit")
+
+  // This should be set in the gatsby-config of the site when enabling this plugin
+  apiServer = options.server
```

Comment:
use `setApiServer` on the class instead of a global variable. Why else have a class? 🙃

Reply:
I'm not in the java camp of "every property must have a getter and setter" so I don't really agree with this.

Since you asked; I use a class as a singleton because that seemed like the best data structure currently available. I'd have gone for a plain object with methods but at that point the class paradigm feels better. (I actually went back and forth between the two)

That said, if you're dead set on having a `getApiServer` and `setApiServer` etc for every property (or if Gatsby does) then I'll add it. Personally I think that if it doesn't do anything else, it just bloats the code and serves no clear purpose over a regular read/writes.

I can't wait for class property syntax to be a thing though.

Reply:
I'm not in the java camp of "every property must have a getter and setter" so I don't really agree with this.

Since you asked; I use a class as a singleton because that seemed like the best data structure currently available. I'd have gone for a plain object with methods but at that point the class paradigm feels better. (I actually went back and forth between the two)

That said, if you're dead set on having a getApiServer and setApiServer etc for every property (or if Gatsby does) then I'll add it. Personally I think that if it doesn't do anything else, it just bloats the code and serves no clear purpose over a regular read/writes.

I can't wait for class property syntax to be a thing though.